src/libmeasurement_kit/dns/system_resolver.hpp

```diff
@@ -0,0 +1,174 @@
+            break;
+        case EAI_OVERFLOW:
+            resolver_error = ArgumentBufferOverflow();
+            break;
+        case EAI_PROTOCOL:
+            resolver_error = UnknownResolvedProtocol();
+            break;
+        case EAI_SERVICE:
+            resolver_error = NotSupportedServname();
+            break;
+        case EAI_SOCKTYPE:
+            resolver_error = NotSupportedAISocktype();
+            break;
+        default:
+            resolver_error = ResolverError();
+        }
+        ctx->reactor->call_soon([=]() { callback(resolver_error, nullptr); });
+        return;
+    }
+    assert(servinfo != nullptr);
+
+    void *addr_ptr;
+    char address[128];
+    for (addrinfo *p = servinfo.get(); p != nullptr; p = p->ai_next) {
+        Answer answer;
+        answer.name = ctx->name;
+        answer.qclass = ctx->dns_class;
+        assert(p->ai_family == AF_INET or p->ai_family == AF_INET6);
+        if (p->ai_family == AF_INET) {
+            answer.type = QueryTypeId::A;
+            addr_ptr = &((sockaddr_in *)p->ai_addr)->sin_addr;
+        } else if (p->ai_family == AF_INET6) {
+            answer.type = QueryTypeId::AAAA;
+            addr_ptr = &((sockaddr_in6 *)p->ai_addr)->sin6_addr;
+        }
+        if (p->ai_canonname != nullptr) {
+            answer.hostname = p->ai_canonname;
+        }
+        if (inet_ntop(p->ai_family, addr_ptr, address, sizeof(address)) ==
+            nullptr) {
+            ctx->logger->warn("dns: unexpected inet_ntop failure");
+            ctx->reactor->call_soon(
+                [=]() { callback(InetNtopFailure(), nullptr); });
+            return;
+        }
+        if (p->ai_family == AF_INET) {
+            answer.ipv4 = std::string(address);
+        } else if (p->ai_family == AF_INET6) {
+            answer.ipv6 = std::string(address);
+        }
+        message->answers.push_back(answer);
+    }
+    ctx->reactor->call_soon([=]() { callback(NoError(), message); });
+}
+
+template <MK_MOCK(getaddrinfo), MK_MOCK(inet_ntop)>
+void system_resolver(QueryClass dns_class, QueryType dns_type, std::string name,
+                     Callback<Error, Var<Message>> cb, Settings settings,
+                     Var<Reactor> reactor, Var<Logger> logger) {
+    std::unique_ptr<ResolverContext> ctx(new ResolverContext(
+        dns_class, dns_type, name, cb, settings, reactor, logger));
+    Query query;
+    ctx->hints.ai_flags = AI_ALL | AI_V4MAPPED;
+    ctx->hints.ai_socktype = SOCK_STREAM;
+
+    if (dns_class != QueryClassId::IN) {
+        reactor->call_soon([=]() { cb(UnsupportedClassError(), nullptr); });
+        return;
+    }
+
+    if (dns_type == QueryTypeId::A) {
+        ctx->hints.ai_family = AF_INET;
+    } else if (dns_type == QueryTypeId::AAAA) {
+        ctx->hints.ai_family = AF_INET6;
+    } else if (dns_type == QueryTypeId::CNAME) {
+        ctx->hints.ai_family = AF_UNSPEC;
+        ctx->hints.ai_flags |= AI_CANONNAME;
```

Comment:
Question: my understanding is that we always want to know the canonical name. Perhaps we should always pass in `AI_CANONNAME`, or do we want the user to specify it explicitly?